code/datums/components/transforming.dm

```diff
@@ -0,0 +1,272 @@
+	var/datum/callback/pre_transform_callback
+	/// Callback to be invoked whenever the weapon is transformed.
+	var/datum/callback/on_transform_callback
+	/// Cooldown in between transforms
+	COOLDOWN_DECLARE(transform_cooldown)
+
+/datum/component/transforming/Initialize(
+		start_transformed = FALSE,
+		transform_cooldown_time = 0 SECONDS,
+		force_on = 0,
+		throwforce_on = 0,
+		throw_speed_on = 2,
+		sharpness_on = NONE,
+		hitsound_on = 'sound/weapons/blade1.ogg',
+		w_class_on = WEIGHT_CLASS_BULKY,
+		clumsy_check = TRUE,
+		list/attack_verb_continuous_on,
+		list/attack_verb_simple_on,
+		datum/callback/pre_transform_callback,
+		datum/callback/on_transform_callback,
+		)
+
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	var/obj/item/item_parent = parent
+
+	src.transform_cooldown_time = transform_cooldown_time
+	src.force_on = force_on
+	src.throwforce_on = throwforce_on
+	src.throw_speed_on = throw_speed_on
+	src.sharpness_on = sharpness_on
+	src.hitsound_on = hitsound_on
+	src.w_class_on = w_class_on
+	src.clumsy_check = clumsy_check
+
+	if(attack_verb_continuous_on)
+		src.attack_verb_continuous_on = attack_verb_continuous_on
+		attack_verb_continuous_off = item_parent.attack_verb_continuous
+	if(attack_verb_simple_on)
+		src.attack_verb_simple_on = attack_verb_simple_on
+		attack_verb_simple_off = item_parent.attack_verb_simple
+
+	src.pre_transform_callback = pre_transform_callback
+	src.on_transform_callback = on_transform_callback
+
+	if(start_transformed)
+		toggle_active(parent)
+
+/datum/component/transforming/RegisterWithParent()
+	var/obj/item/item_parent = parent
+
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK_SELF, .proc/on_attack_self)
+	if(item_parent.sharpness || sharpness_on)
+		RegisterSignal(parent, COMSIG_ITEM_SHARPEN_ACT, .proc/on_sharpen)
+
+/datum/component/transforming/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_ITEM_ATTACK_SELF, COMSIG_ITEM_SHARPEN_ACT))
+
+/datum/component/transforming/Destroy()
+	if(pre_transform_callback)
+		QDEL_NULL(pre_transform_callback)
+	if(on_transform_callback)
+		QDEL_NULL(on_transform_callback)
+	return ..()
+
+/*
+ * Called on [COMSIG_ITEM_ATTACK_SELF].
+ *
+ * Check if we can transform our weapon, and if so, call [do_transform].
+ * (If we have a [pre_transform_callback], invoke it. If that callback returns a falsy value, cancel the transform.)
+ * And, if [do_transform] was successful, do a clumsy effect from [clumsy_transform_effect].
+ *
+ * source - source of the signal, the item being transformed / parent
+ * user - the mob transforming the weapon
+ */
+/datum/component/transforming/proc/on_attack_self(obj/item/source, mob/user)
+	SIGNAL_HANDLER
+
+	if(!COOLDOWN_FINISHED(src, transform_cooldown))
+		to_chat(user, span_warning("Wait a bit before trying to use [source] again!"))
+		return
+
+	if(pre_transform_callback)
+		if(!pre_transform_callback.Invoke(user))
+			return
```

Comment:
You could do this with signals, but this is alright too, just a bit more limiting

Reply:
Sure